server/app/routes/v1/volumes_api.rb

```diff
@@ -0,0 +1,113 @@
+module V1
+  class VolumesApi < Roda
+    include TokenAuthenticationHelper
+    include CurrentUser
+    include RequestHelpers
+    include Auditor
+    include LogsHelpers
+
+    plugin :backtracking_array
+
+    route do |r|
+
+      validate_access_token
+      require_current_user
+
+      def find_grid(grid_name)
+        grid = Grid.find_by(name: grid_name)
+        halt_request(404, {error: "Grid #{grid_name} not found"}) if !grid
+        grid
+      end
+
+      def find_stack(stack_name, grid)
+        stack = grid.stacks.find_by(name: stack_name)
+        halt_request(404, {error: "Stack #{stack_name} not found"}) if !stack
+        stack
+      end
+
+      # @param [String] grid_name
+      # @param [String] stack_name
+      # @param [String] volume_name
+      # @return [GridService]
+      def load_volume(grid_name, stack_name, volume_name)
+        grid = find_grid(grid_name)
+        if stack_name
+          stack = find_stack(stack_name, grid)
+          volume = stack.volumes.find_by(name: volume_name)
+        else
+          volume = grid.volumes.find_by(name: volume_name)
+        end
+        halt_request(404, {error: "Volume #{volume_name} not found"}) if !volume
+
+        unless current_user.grid_ids.include?(volume.grid_id)
+          halt_request(403, {error: 'Access denied'})
+        end
+
+        volume
+      end
+
+      # Parse stack/vol out of path
+      # @param [String] volume path
+      # @return [Array] [stack, volume]
+      def parse_params(path)
```

Comment:
This is weird... alternative would be the same `:grid/null/:volume` vs `:grid/:stack/:volume` as used for the services API. Keep the `[nil, volume]` return value, though :)

Mainly this kind of schema prevents ever adding new `POST :grid/:volume/snapshot` API endpoints.